### ISSUE TITLE
lws_callback_on_writable does not initiate a callback

### ISSUE THREAD

**Issue by tsarthak** [warmcat/libwebsockets#2510]
Hello Andy, team!

I have recently started developing with LWS and this perhaps might be a really stupid question, but I am struggling to get `WRITABLE_` callbacks on lws.
The objective is to 
1. Connect to a remote server over WS (it works)
2. When the connection is established invoke a callback to my function which sends data using the below function, send some auth data and get back some tokens under `CLIENT_RECEIVE` (it works)
3. Write to this same WS in case client wants some information from the server at a later point. using the service loop, and invoking lws_callback_on_writable(does not work).
4. Keep this WS open and receive some data from the server in the future. (don't know if this works)

To better describe what I am doing, here are some relevant snippets.

```cpp
// send some data
void send_data(std::vector<uint8_t> &&data, data_traits trait = data_traits::text)
        {
            m_data_queue.push(std::make_pair(trait, std::move(data))); // this data is already PRE_ PADDED
            lws_callback_on_writable(m_client_wsi); // trying to invoke the callback
        }

// 
static int event_callback(struct lws *wsi, enum lws_callback_reasons reason, void *user, void *in, size_t len)
    {
        switch (reason)
        {
            /* because we are protocols[0] ... */
            case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
            {
                std::stringstream error_ss;
                error_ss << "Client connection error: " << (in ? static_cast<char *>(in) : "UNKNOWN");
                if (_p->on_error)
                    _p->on_error("");
                _p->m_client_wsi = nullptr;
                _p->m_connected = false;
                break;
            }
            case LWS_CALLBACK_CLIENT_ESTABLISHED:
            {
                _p->m_connected = true;
                if (_p->on_connect)
                    _p->on_connect();
                break;
            }
            case LWS_CALLBACK_CLIENT_RECEIVE:
            {
                auto casted = static_cast<uint8_t *>(in);
                std::vector<uint8_t> data(casted, casted + len);
                if (_p->on_receive)
                    _p->on_receive(data);
                break;
            }
            case LWS_CALLBACK_CLIENT_CLOSED:
            {
                _p->m_client_wsi = nullptr;
                _p->m_connected = false;
                if (_p->on_disconnect)
                    _p->on_disconnect();
                break;
            }
            case LWS_CALLBACK_CLIENT_WRITEABLE:
            {
                    auto data_with_trait = std::move(_p->m_data_queue.pop());
                    int ret_val = lws_write(wsi, data_with_trait.second.data(), data_with_trait.second.size(), LWS_WRITE_TEXT);
                    if (_p->on_send)
                        _p->on_send(ret_val);

                    if (!_p->m_data_queue.empty())
                        lws_callback_on_writable(_p->m_client_wsi); // trying to re-invoke the callback for pending data

                break;
            }
            default:
                break;
        }
        return lws_callback_http_dummy(wsi, reason, user, in, len);
    }
```

Having read some of the past issues here, it seems that it is NOT guaranteed to work. i.e. n calls to `lws_callback_on_writable` do not guarantee n callbacks, based on the status of the underlying connection if the peer is ready to accept more data.
In that case I have a few questions
1. Is it possible to force this callback? I know it is not advisable and this will not end up in production, but would be good to know from a testing POV sometimes.
2. Is it possible to inspect what is the status of the stream? Or perhaps find out WHY is the stream not ready to accept data? Because AFAIK, the server does nothing special besides writing back some response. Is there something that I need to handle when the client receives/writes something to make sure that the stream is writable in the future?
3. Is there a better way to do this with LWS that I have completely glossed over?

I open to suggestions and look forward to your response!

I am using LWS 4.3.0 stable on macOS 12 using C++17 with clang/lldb and OpenSSL and MbedTLS for TLS.

**Comment by lws-team** [comment]
When people turn up with this symptom, it's usually because they don't realize the relationship between event loops and thread safety.

https://libwebsockets.org/git/libwebsockets/tree/READMEs/README.event-loops-intro.md

> force

Anything's possible, but that isn't going to help if your problem is thinking lws is threadsafe.

> inspect

POSIX sockets don't provide this kind of information.  Generally writeability is delayed when there is too much already in flight for the tcp window, it will become writeable when more ACKs are received.  So it regulates outgoing data rate this way.  If you wrote something bigger than the OS would accept, writeable callbacks are concealed from your code while lws finishes sending your oversize data in the background.  On muxed connections like h2, streams are allowed to write in round-robin order and may have to wait for others to go first.

Otherwise, the callback will happen reliably the next time around the event loop.  The thing about not getting n callbacks for n requests is you are setting a flag, once it's set, setting it again won't do anything different until you got a callback and your flag was reset.  You can also get additional callbacks coming under some conditions, so the cb should decide what to write (if anything) based on its knowledge of the stream state.

**Comment by tsarthak** [comment]
Hi Andy, Thank you for such a swift response!
You were spot on and I understand the problem. Finally. Here is what helped - 

> Lws apis cannot be called safely from any thread other than the event loop one,
> with the sole exception of lws_cancel_service().

The problem was the `lws_callback_on_writeable` requires a call from the service loop thread (which is separate). It works well now.

Thank you for your help and for reaching out so soon

Merry Christmas and a happy new year to you and the team!

**Comment by pvouzis** [comment]
@lws-team I noticed that on another issue (2017) https://github.com/warmcat/libwebsockets/issues/1125 you mention that `lws_cancel_service()` and `lws_callback_on_writable()` can be called by another thread. But here (https://libwebsockets.org/git/libwebsockets/tree/READMEs/README.event-loops-intro.md) I read that only `lws_cancel_service()` can be called by another thread.

I wanted to double check if things changed since 2017 and confirm that `lws_callback_on_writable()` is not thread safe.

**Comment by lws-team** [comment]
Yes although I intended it to be like that, and I did a lot of work to enable it (eg, BSD poll() overwrites changes to the list of fds we are waiting on when it wakes, so losing any changes done by the other thread; I implemented workarounds for it), it's basically on a loser because lws has to support event libraries with the same api.

It's opaque what wait they actually use in the library and it's a loser to second-guess it or the event library's internal reaction to having its apis called from other threads.  For that reason it had to step back that the only guarantee I could make was for lws_cancel_service, where I can control how it works.
